commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/VariableSizeChunkIndex.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2023 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tiered.storage.commons.chunkindex;
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * The chunk index for transformed chunks of variable size.
+ *
+ * <p>The most important quality is that while the original file chunks
+ * are of the same size (apart from maybe the final one), the transformed chunks
+ * have different sizes that must be specified and stored explicitly.
+ * For example, this is a result of compression.
+ *
+ * <p>An example:
+ * <pre>
+ * Original file:        |  Transformed file:
+ * [0-100)   - size 100  |  [0-30)  - size 30
+ * [100-200) - size 100  |  [30-50) - size 20
+ * [200-250) - size 50   |  [50-60) - size 10
+ * </pre>
+ */
+public class VariableSizeChunkIndex extends AbstractChunkIndex {
+    private final List<Integer> transformedChunks;
+
+    // This only a materialization for convenience and performance,
+    // it should not be persisted.
+    private final List<Chunk> chunks;
+
+    public VariableSizeChunkIndex(final int originalChunkSize,
+                                  final int originalFileSize,
+                                  final List<Integer> transformedChunks) {
+        super(originalChunkSize, originalFileSize, transformedChunks.get(transformedChunks.size() - 1));
+        this.transformedChunks = List.copyOf(Objects.requireNonNull(transformedChunks));
+
+        chunks = materializeChunks();
+    }
+
+    public List<Chunk> chunks() {
+        return chunks;
```

Comment:
Instead of copying the list in the constructor I guess we would want to return a copy here to keep this list unmodifiable.

Reply:
I mean, `materializeChunks` produce already an unmodifiable list

Reply:
Oh, sorry I was thinking about different field when mentioned constructor.
Regarding `materializeChunks` looks like I'm missing some context here. Is the idea here also to cache the produced list as a field in the instance of this object and that is the reason why we are calling `materializeChunks` from constructor and not from here directly? And moreover this *Index classes supposed to be immutable, correct?(This I guess should be mentioned somewhere for clarity)

Reply:
> Is the idea here also to cache the produced list as a field in the instance of this object and that is the reason why we are calling `materializeChunks` from constructor and not from here directly?

Yeah, the idea is to pre-calculate the chunks list and cache it. It makes sense because the chunk list is inevitable to be used and potentially multiple times. It's done in the constructor, because it's automatically thread-safe this way (no need to have locks/`volatile`).

> And moreover this *Index classes supposed to be immutable, correct?(This I guess should be mentioned somewhere for clarity)

Yes, you're right. I added this clarification.

Reply:
> >Is the idea here also to cache the produced list as a field in the instance of this object and that is the reason why we are calling materializeChunks from constructor and not from here directly?
>
> Yeah, the idea is to pre-calculate the chunks list and cache it. It makes sense because the chunk list is inevitable to be used and potentially multiple times. It's done in the constructor, because it's automatically thread-safe this way (no need to have locks/volatile).

I'm starting to hesitate that it's a right approach taking into account all potential initialization issues. Probably we could use a similar approach but move this `materializeChunks` out of this hierarchy to some kind of wrapper maybe..

Reply:
Extracting it in any form will not work as the algorithm inextricably depends on the object's state: in one way or another we'll have to pass closures on the state. The only way to make it "right" is to make two copies of `transformedChunkSize` and `originalChunkSize`, one to be called in the constructor and the other in other places, which doesn't make sense.

The book's advice is generally right, but it in this particular case there is no issue. The present code execution is correct and we left the warnings for the posterity in case the classes are to be modified or extended. I don't see how realistically it can fail.

Reply:
Revisited it once again and with additions of finals seems to be fine now.